### DISCUSSION TITLE
Atom Attributes

### DISCUSSION THREAD

**Discussion by rbdavid** [BradyAJohnston/MolecularNodes#396] [General]
I'm inspecting the possible attributes that I can utilize for visualizing protein structures. Previously, I've used both the "b_factor" and "occupancy" columns in .pdb/.cif files to store calculated values. Then, in VMD, I visualized these metrics on the structure by coloring the cartoon or licorice reps by the respective column's values. 

I'm noticing a couple issues that are breaking this visualization habit I've got:
- The occupancy field/attribute is never created.
- ~~The b_factor attribute value for each atom is set as the b_factor value for the residue's CA atom (for proteins). I haven't looked at how nucleic polymers, ligands, and solvent are treated.~~ Was looking at the "Evaluated" spreadsheet tab instead of "original".

Both issues happen whether you download a structure or load from a local file. Ideally, when a structure file is imported into a blender object, each atom would have its own b_factor and occupancy values, as reported in the structure file. Considering experimental structures, both columns hold important information about each atom in the structural model. Side chain atoms may or may not have higher b_factor values, which is insightful information about the underlying ensemble of sidechain conformations present in the crystal. Similarly, occupancy values less than 1 indicate that the structural modeling software determined there were a number of equivalent conformations for a residue. 

Alternatively, how easy is it to import data files, create named attributes for each atom in a system, and then use those attributes in the geometry nodes for visualizing the structure?

**Comment by rbdavid** [comment]
https://www.youtube.com/watch?v=wrLZ6SIE_7c shows an add-on to read in a csv file and subsequently add the file's columns as attributes to a blender object. This seems like the easiest gui-based method of creating attributes and then using them in the geometry nodes. 

Ideally, this process can be automated using the blender python console. So I need to write a script that reads in a .csv file and adds the attributes to the atoms (the mesh vertices of the structure's blender object). Also, I've gotta create the geo nodes needed to visualize the attribute. 

From then on, I can `Append` that node group to each new blender project, run the script for the relevant .csv file, and set up the nodes to get the visualization that I want. Simple as.

**Reply by BradyAJohnston** [reply]
Yes you can in theory import any arbitrary values and assign them to the mesh. You'll need the length of the data to exactly match the number of vertices.  You can do so using MN: 

```python
import molecularnodes as mn

mn.io.blender.obj.add_attribute(object, 'some_attribute', data_array)
```

I haven't settled on a stable python API (and I welcome feedback in #359 on it), so this will likely change in the future, but this is where the function is currently defined: 
https://github.com/BradyAJohnston/MolecularNodes/blob/5d41abe60f1f7de055700c83ec18c46e7b321810/molecularnodes/blender/obj.py#L68-L113

An important gotcha when manually writing from `numpy` to meshes, is the `.copy(order='c')` otherwise Blender will crash, and when writing a 2D array you need to reshape it to a 1D array `.reshape(-1)` again otherwise Blender won't like it. Even if you are writing to a vector attribute the data can only be written as a 1D array.

**Comment by rbdavid** [comment]
In the `Scripting` workspace, you can find template scripts (`Templates`->`Python`->`Bmesh Simple`) for doing a basic transformation on an object's mesh vertices.  And then https://blender.stackexchange.com/questions/266991/manually-setting-custom-attributes-per-edge-vertex-etc has some examples of adding attributes. All that I need to do then is write the csv reader using modules available within the Blender python interpreter; I'm thinking pandas or csv. Doesn't really matter.

**Comment by rbdavid** [comment]
Oh son of a... I'm looking at the wrong tab of the spread sheet. So the atoms' b_factor columns do include the correct values. grumble grumble grumble.

**Reply by BradyAJohnston** [reply]
I have additionally just added occupancy to also be imported in #398 . Will make a release including the updates later today
